botocore/handlers.py

```diff
@@ -51,6 +52,7 @@
 # combination of uppercase letters, lowercase letters, numbers, periods
 # (.), hyphens (-), and underscores (_).
 VALID_BUCKET = re.compile('^[a-zA-Z0-9.\-_]{1,255}$')
+VERSION_ID_SUFFIX = re.compile(r'\?versionId=[^\s]+$')
```

Comment:
Not opposed.  The main reason I didn't do this is because S3 will return validation errors if you have anything other than versionId in your query string.  It's possible this can change in the future on S3's side, but with the proposed change of supporting both string and dict, hopefully this won't matter as we can steer people towards using the dict format.